components/style/selector_parser.rs

```diff
@@ -173,3 +173,25 @@ impl<T> PerPseudoElementMap<T> {
         self.entries.iter()
     }
 }
+
+/// Values for the :dir() pseudo class
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Direction {
+    /// left-to-right semantic directionality
+    Ltr,
+    /// right-to-left semantic directionality
+    Rtl,
+    /// Some other provided directionality value
+    Other(Box<str>),
+}
+
+impl ToCss for Direction {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: Write {
+        let dir_str = match *self {
+            Direction::Rtl => "rtl",
+            Direction::Ltr => "ltr",
+            Direction::Other(ref other) => other,
```

Comment:
Please add a `FIXME` here saying:

> FIXME: This should be escaped as an identifier.

That can be a followup if you want, since that's a bugfix, requires tests and all that.

The short version is that it should use `serialize_identifier`, in order to properly handle escapes. I believe we have tests that test this where you could grab an example... But looking at what `serialize_identifier` handles, something like this:

```css
:dir(\-) { .. }
```

Should be a valid rule, but serializes back to:

```css
:dir(-) { .. }
```

Which isn't an identifier token IIRC.

Reply:
Created #19231 which I can pick up as a followup.